iib/workers/tasks/build.py

```diff
@@ -1420,7 +1420,7 @@ def _apply_package_name_suffix(metadata_path, organization=None):
     with open(annotations_yaml_path, 'r') as f:
         try:
             annotations_yaml = yaml.load(f)
-        except ruamel.yaml.YAMLError:
```

Comment:
I'm puzzled here, this is inheritance:

```
Exception --> YAMLError --> MarkedYAMLError --> ScannerError(MarkedYAMLError)
```
https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/scanner.py
https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/error.py

So scanner error should be already covered by `YAMLError`

Reply:
https://github.com/release-engineering/iib/blob/master/iib/workers/tasks/build.py#L1501  This is the line that causes issues

Reply:
Thanks!

Reply:
I don't think this is the right place. From the traceback:
```
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,624: DEBUG/ForkPoolWorker-1] The "package_name_suffix" configuration is not set for the organization certified-operators
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,719: INFO/ForkPoolWorker-1] Setting the state of request 19106 to "failed" with the reason "An unknown error occurred"
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,719: INFO/ForkPoolWorker-1] Patching the request 19106 with {'state': 'failed', 'state_reason': 'An unknown error occurred'}
Oct 08 13:43:00 iib-main celery[2130]: [2020-10-08 13:43:00,512: ERROR/ForkPoolWorker-1] Task iib.workers.tasks.build.handle_regenerate_bundle_request[51f9e164-4f84-466c-8f51-176018d91c11] >
Oct 08 13:43:00 iib-main celery[2130]: Traceback (most recent call last):
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/celery/app/trace.py", line 385, in trace_task
Oct 08 13:43:00 iib-main celery[2130]:     R = retval = fun(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/celery/app/trace.py", line 650, in __protected_call__
Oct 08 13:43:00 iib-main celery[2130]:     return self.run(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/utils.py", line 263, in wrapper
Oct 08 13:43:00 iib-main celery[2130]:     return func(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/build.py", line 1244, in handle_regenerate_bundle_request
Oct 08 13:43:00 iib-main celery[2130]:     manifests_path, metadata_path, organization, pinned_by_iib
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/build.py", line 1425, in _adjust_operator_bundle
Oct 08 13:43:00 iib-main celery[2130]:     operator_manifest = OperatorManifest.from_directory(manifests_path)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 555, in from_directory
Oct 08 13:43:00 iib-main celery[2130]:     operator_csvs = list(cls._get_csvs(yaml_files, **kwargs))
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 569, in _get_csvs
Oct 08 13:43:00 iib-main celery[2130]:     yield OperatorCSV.from_file(f, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 317, in from_file
Oct 08 13:43:00 iib-main celery[2130]:     data = yaml.load(f)
...
...
...
Oct 08 13:43:00 iib-main celery[2130]: ruamel.yaml.scanner.ScannerError: while scanning for the next token
Oct 08 13:43:00 iib-main celery[2130]: found character '\t' that cannot start any token
Oct 08 13:43:00 iib-main celery[2130]:   in "/tmp/iib-swyr59vk/manifests/cic-operator.v1.7.6.clusterserviceversion.yaml", line 195, column 9
```

That looks like this line: https://github.com/release-engineering/iib/blob/aec253346d6c51cb52293154f5008dc19048f079/iib/workers/tasks/build.py#L1501
The line number is off, but that's just because the code base has changed since October 8th.

It also matches my understanding of the issue. The error is seen when there are issues in the CSV file which defines the operator manifest, not the annotations YAML file.

I think it's important to catch the `ruamel.yaml.constructor.DuplicateKeyError` exception and provide a meaningful error message just because we see this so often in IIB. In reality, there could be many other errors, not all related to invalid YAML. 

The problem with not catching errors is that the traceback is logged on the server side and the user has no idea why the IIB build failed. There's something simple we can do this to save ourselves some headache in the future. For all IIB celery tasks, this error callback is installed, https://github.com/release-engineering/iib/blob/master/iib/workers/tasks/general.py#L14

We can leverage this to ensure any error is logged properly and visible to the user. To do this, you can add the `request_logger` decorator to that task, similarly to how it's done to the other celery tasks in build.py, for example. Then, a simple call to `log.error(msg, exc_info=exc)` within `failed_request_callback` will make sure the error is propagated back to the user.

It would be great to do both of those things:
1. Provide a meaningful error for duplicated keys in the operator manifest, and
2. Always log errors in the build specific log file